keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+`Node.Status.Hostname` and `Node.Status.IPs`, with clearer semantics
+and more consistent and useful behavior.
+
+## Motivation
+
+### Problems
+
+#### IPv6 and Dual-stack
+
+Nodes implicitly have a "primary" node IP (the first `InternalIP`
+address, unless there are no `InternalIP` addresses, in which case the
+first `ExternalIP` address) which is used for a handful of purposes,
+such as the `Pod.Status.HostIP` value, and by extension, the
+`Pod.Status.PodIP` of a host-network pod. This implies that
+single-stack clusters must have a primary node IP of the same family
+that the pod network uses, or else pod-network pods would be unable to
+communicate with host-network pods, violating [the Kubernetes network
+model].
+
+When using an external cloud provider, cloud-controller-manager is
+responsible for setting `Node.Status.Addresses`, but it does not know
+what IP families pods are using, so it doesn't know whether to return
+an IPv4 or IPv6 address first as the primary IP. (In theory this
+problem could be solved by letting it return them in any order, and
+changing kubelet's definition of what the primary IP is. In practice
+this would not work well since some external components also use the
+same "first `InternalIP` ..." rule that kubelet uses now.)
+
+In a dual-stack cluster, there is currently no definition of
+"secondary" / "alternate-family primary" node IP, and host-network
+pods currently only get a single IP in `Pod.Status.PodIPs`, which in
+turn means that in any given cluster, host-network pods can either
+only be the endpoints of IPv4 Services, or else only be the endpoints
+of IPv6 Services.
+
+In general, it is not clear if `Node.Status.Addresses` is
+expected/allowed to contain IPv6 addresses in a single-stack IPv4
+cluster, or IPv4 addresses in a single-stack IPv6 cluster. For cloud
+providers that do not currently return IPv6 addresses, starting to
+return them might break clients which were written without taking this
+possibility into account.
+
+[the Kubernetes network model]: https://kubernetes.io/docs/concepts/cluster-administration/networking/#the-kubernetes-network-model
+
+#### Address Detection Behavior Across Different Cluster Types
+
+On bare-metal, a node always has exactly one IP (even if the cluster
+is supposed to be dual-stack), which is auto-detected by kubelet, but
+can be overridden with `--node-ip`.
+
+When using a legacy cloud provider, the node IPs are detected by the
+cloud provider, and it may return multiple IPs of each type (eg, if
+the node has multiple network interfaces, or multiple addresses on a
+single interface). eg:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.1.2.3" },
+      { "type": "InternalIP", "address": "10.4.5.6" },
+      { "type": "InternalIP", "address": "10.7.8.9" } ]
+
+In this case, the primary IP would be `10.1.2.3`. If the administrator
+wants a different primary IP, they can pass `--node-ip` to choose
```

Comment:
I assume you mean passing `--node-ip` to the kubelet, but it might be good to make the explicit.